src/mame/casio/wk1800.cpp

```diff
@@ -0,0 +1,615 @@
+void wk1800_state::wk1800_map(address_map& map)
+{
+	map(0x00000, 0x1ffff).rom();
+	map(0x20000, 0x2ffff).rw(m_gt155, FUNC(gt155_device::read), FUNC(gt155_device::write));
+	map(0x30000, 0x30001).mirror(0x0fff0).r("kbd", FUNC(gt913_kbd_hle_device::read));
+	map(0x30002, 0x30003).mirror(0x0fff0).r("kbd", FUNC(gt913_kbd_hle_device::status_r));
+	map(0x40000, 0x40003).mirror(0x1fffc).m(m_fdc, FUNC(hd63266f_device::map));
+	map(0x60000, 0x7ffff).rw(m_fdc, FUNC(hd63266f_device::dma_r), FUNC(hd63266f_device::dma_w));
+	map(0x80000, 0xbffff).mirror(0x40000).ram().share("nvram");
+}
+
+
+/**************************************************************************/
+void wk1800_state::driver_start()
+{
+	m_led.resolve();
+	m_led_power.resolve();
+	m_outputs.resolve();
+
+	m_nmi_timer = timer_alloc(FUNC(wk1800_state::nmi_clear), this);
+
+	std::fill(std::begin(m_sound_regs), std::end(m_sound_regs), 0);
+	std::fill(std::begin(m_dsp_data), std::end(m_dsp_data), 0);
+	m_sound_rom_addr = 0;
+
+	m_led_sel = 0xff;
+	m_input_sel = 0;
+
+	m_led_clk = m_input_clk = m_shift_data = 0;
+
+	save_item(NAME(m_sound_regs));
+	save_item(NAME(m_dsp_data));
+	save_item(NAME(m_sound_rom_addr));
+
+	save_item(NAME(m_led_sel));
+	save_item(NAME(m_input_sel));
+
+	save_item(NAME(m_led_clk));
+	save_item(NAME(m_input_clk));
+	save_item(NAME(m_shift_data));
+}
+
+
+/**************************************************************************/
+void wk1800_state::wk1600(machine_config &config)
+{
+	H83048(config, m_maincpu, 16'000'000).set_mode_a20();
+	m_maincpu->set_addrmap(AS_PROGRAM, &wk1800_state::wk1600_map);
+	m_maincpu->read_adc<0>().set_constant(0);
+	m_maincpu->read_adc<1>().set_ioport("AN1");
+	m_maincpu->read_adc<2>().set_constant(0);
+	m_maincpu->read_adc<3>().set_ioport("AN3");
+	m_maincpu->read_port6().set_ioport("P6");
+	m_maincpu->read_port7().set_ioport("P7");
+	m_maincpu->read_port8().set_ioport("P8");
+	m_maincpu->write_port8().set_ioport("P8");
+	m_maincpu->read_port9().set_ioport("P9");
+	m_maincpu->read_porta().set_ioport("PA");
+	m_maincpu->write_porta().set_ioport("PA");
+	m_maincpu->read_portb().set(FUNC(wk1800_state::lcd_r));
+	m_maincpu->write_portb().set_ioport("PB");
+
+	NVRAM(config, "nvram");
+
+	GT913_KBD_HLE(config, "kbd"); // actually TC190C020AF-001 gate array
+
+	auto& mdin(MIDI_PORT(config, "mdin"));
+	midiin_slot(mdin);
+	mdin.rxd_handler().set(m_maincpu, FUNC(h83048_device::sci_rx_w<0>));
+
+	auto& mdout(MIDI_PORT(config, "mdout"));
+	midiout_slot(mdout);
+	m_maincpu->write_sci_tx<0>().set(mdout, FUNC(midi_port_device::write_txd));
+
+	HD44780(config, m_lcdc, 270'000); // TODO: Wrong device type, should be SED1278F2A; clock not measured, datasheet typical clock used
+	m_lcdc->set_lcd_size(2, 8);
+
+	auto& screen = SCREEN(config, "screen", SCREEN_TYPE_SVG);
+	screen.set_refresh_hz(60);
+	screen.set_size(1755, 450);
+	screen.set_visarea_full();
+	screen.screen_vblank().set(FUNC(wk1800_state::render_w));
```

Comment:
Can we keep the locals for instantiated devices consistent?  The screen one is the odd one out here.  Most of MAME tends to do:
```c++
	auto &screen(SCREEN(config, "screen", SCREEN_TYPE_SVG));
```
for these.